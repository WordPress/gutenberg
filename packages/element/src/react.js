/**
 * External dependencies
 */
// eslint-disable-next-line @typescript-eslint/no-restricted-imports
import {
	Children,
	cloneElement,
	Component,
	createContext,
	createElement,
	createRef,
	forwardRef,
	Fragment,
	isValidElement,
	memo,
	PureComponent,
	StrictMode,
	useCallback,
	useContext,
	useDebugValue,
	useDeferredValue,
	useEffect,
	useId,
	useMemo,
	useImperativeHandle,
	useInsertionEffect,
	useLayoutEffect,
	useReducer,
	useRef,
	useState,
	useSyncExternalStore,
	useTransition,
	startTransition,
	lazy,
	Suspense,
} from 'react';

/**
 * Object containing a React element.
 *
 * @typedef {import('react').ReactElement} Element
 */

/**
 * Object containing a React component.
 *
 * @typedef {import('react').ComponentType} ComponentType
 */

/**
 * Object containing a React synthetic event.
 *
 * @typedef {import('react').SyntheticEvent} SyntheticEvent
 */

/**
 * Object containing a React ref object.
 *
 * @template T
 * @typedef {import('react').RefObject<T>} RefObject<T>
 */

/**
 * Object containing a React ref callback.
 *
 * @template T
 * @typedef {import('react').RefCallback<T>} RefCallback<T>
 */

/**
 * Object containing a React ref.
 *
 * @template T
 * @typedef {import('react').Ref<T>} Ref<T>
 */

/**
 * Object that provides utilities for dealing with React children.
 */
export { Children };

/**
 * Creates a copy of an element with extended props.
 *
 * @param {Element} element Element
 * @param {?Object} props   Props to apply to cloned element
 *
 * @return {Element} Cloned element.
 */
export { cloneElement };

/**
 * A base class to create WordPress Components (Refs, state and lifecycle hooks)
 */
export { Component };

/**
 * Creates a context object containing two components: a provider and consumer.
 *
 * @param {Object} defaultValue A default data stored in the context.
 *
 * @return {Object} Context object.
 */
export { createContext };

/**
 * Returns a new element of given type. Type can be either a string tag name or
 * another function which itself returns an element.
 *
 * @param {?(string|Function)} type     Tag name or element creator
 * @param {Object}             props    Element properties, either attribute
 *                                      set to apply to DOM node or values to
 *                                      pass through to element creator
 * @param {...Element}         children Descendant elements
 *
 * @return {Element} Element.
 */
export { createElement };

/**
 * Returns an object tracking a reference to a rendered element via its
 * `current` property as either a DOMElement or Element, dependent upon the
 * type of element rendered with the ref attribute.
 *
 * @return {Object} Ref object.
 */
export { createRef };

/**
 * Component enhancer used to enable passing a ref to its wrapped component.
 * Pass a function argument which receives `props` and `ref` as its arguments,
 * returning an element using the forwarded ref. The return value is a new
 * component which forwards its ref.
 *
 * @param {Function} forwarder Function passed `props` and `ref`, expected to
 *                             return an element.
 *
 * @return {Component} Enhanced component.
 */
export { forwardRef };

/**
 * A component which renders its children without any wrapping element.
 */
export { Fragment };

/**
 * Checks if an object is a valid React Element.
 *
 * @param {Object} objectToCheck The object to be checked.
 *
 * @return {boolean} true if objectToTest is a valid React Element and false otherwise.
 */
export { isValidElement };

/**
 * @see https://react.dev/reference/react/memo
 */
export { memo };

/**
 * Component that activates additional checks and warnings for its descendants.
 */
export { StrictMode };

/**
 * @see https://react.dev/reference/react/useCallback
 */
export { useCallback };

/**
 * @see https://react.dev/reference/react/useContext
 */
export { useContext };

/**
 * @see https://react.dev/reference/react/useDebugValue
 */
export { useDebugValue };

/**
 * @see https://react.dev/reference/react/useDeferredValue
 */
export { useDeferredValue };

/**
 * @see https://react.dev/reference/react/useEffect
 */
export { useEffect };

/**
 * @see https://react.dev/reference/react/useId
 */
export { useId };

/**
 * @see https://react.dev/reference/react/useImperativeHandle
 */
export { useImperativeHandle };

/**
 * @see https://react.dev/reference/react/useInsertionEffect
 */
export { useInsertionEffect };

/**
 * @see https://react.dev/reference/react/useLayoutEffect
 */
export { useLayoutEffect };

/**
 * @see https://react.dev/reference/react/useMemo
 */
export { useMemo };

/**
 * @see https://react.dev/reference/react/useReducer
 */
export { useReducer };

/**
 * @see https://react.dev/reference/react/useRef
 */
export { useRef };

/**
 * @see https://react.dev/reference/react/useState
 */
export { useState };

/**
 * @see https://react.dev/reference/react/useSyncExternalStore
 */
export { useSyncExternalStore };

/**
 * @see https://react.dev/reference/react/useTransition
 */
export { useTransition };

/**
 * @see https://react.dev/reference/react/startTransition
 */
export { startTransition };

/**
 * @see https://react.dev/reference/react/lazy
 */
export { lazy };

/**
 * @see https://react.dev/reference/react/Suspense
 */
export { Suspense };

/**
 * @see https://react.dev/reference/react/PureComponent
 */
export { PureComponent };

/**
 * Concatenate two or more React children objects.
 *
 * @param {...?Object} childrenArguments Array of children arguments (array of arrays/strings/objects) to concatenate.
 *
 * @return {Array} The concatenated value.
 */
export function concatChildren( ...childrenArguments ) {
	return childrenArguments.reduce( ( accumulator, children, i ) => {
		Children.forEach( children, ( child, j ) => {
			if ( child && 'string' !== typeof child ) {
				child = cloneElement( child, {
					key: [ i, j ].join(),
				} );
			}

			accumulator.push( child );
		} );

		return accumulator;
	}, [] );
}

/**
 * Switches the nodeName of all the elements in the children object.
 *
 * @param {?Object} children Children object.
 * @param {string}  nodeName Node name.
 *
 * @return {?Object} The updated children object.
 */
export function switchChildrenNodeName( children, nodeName ) {
	return (
		children &&
		Children.map( children, ( elt, index ) => {
			if ( typeof elt?.valueOf() === 'string' ) {
				return createElement( nodeName, { key: index }, elt );
			}
			const { children: childrenProp, ...props } = elt.props;
			return createElement(
				nodeName,
				{ key: index, ...props },
				childrenProp
			);
		} )
	);
}
