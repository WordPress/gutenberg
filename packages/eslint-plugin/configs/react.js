module.exports = {
	extends: [ 'plugin:react/recommended' ],
	parserOptions: {
		ecmaFeatures: {
			jsx: true,
		},
	},
	settings: {
		react: {
			version: 'detect',
		},
	},
	plugins: [ '@wordpress', 'react', 'react-hooks' ],
	rules: {
		'@wordpress/no-unused-vars-before-return': [
			'error',
			{
				excludePattern: '^use',
			},
		],
		'react/display-name': 'off',
		'react/jsx-curly-spacing': [
			'error',
			{
				when: 'always',
				children: true,
			},
		],
		'react/jsx-equals-spacing': 'error',
		'react/jsx-indent': [ 'error', 'tab' ],
		'react/jsx-indent-props': [ 'error', 'tab' ],
		'react/jsx-key': 'error',
		'react/jsx-tag-spacing': 'error',
		'react/no-children-prop': 'off',
		'react/prop-types': 'off',
		'react/react-in-jsx-scope': 'off',
		'react-hooks/exhaustive-deps': [
			'warn',
			{
				additionalHooks: '^(useSelect|useSuspenseSelect)$',
			},
		],
		'react-hooks/rules-of-hooks': 'error',
	},
};
