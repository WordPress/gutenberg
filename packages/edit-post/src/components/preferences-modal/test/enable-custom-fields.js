/**
 * External dependencies
 */
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

/**
 * Internal dependencies
 */
import {
	EnableCustomFieldsOption,
	CustomFieldsConfirmation,
} from '../enable-custom-fields';

describe( 'EnableCustomFieldsOption', () => {
	it( 'renders a checked checkbox when custom fields are enabled', () => {
		const { container } = render(
			<EnableCustomFieldsOption areCustomFieldsEnabled />
		);

		expect( container ).toMatchSnapshot();
	} );

	it( 'renders an unchecked checkbox when custom fields are disabled', () => {
		const { container } = render(
			<EnableCustomFieldsOption areCustomFieldsEnabled={ false } />
		);

		expect( container ).toMatchSnapshot();
	} );

	it( 'renders an unchecked checkbox and a confirmation message when toggled off', async () => {
		const user = userEvent.setup();

		const { container } = render(
			<EnableCustomFieldsOption areCustomFieldsEnabled />
		);

		await user.click( screen.getByRole( 'checkbox' ) );

		expect( container ).toMatchSnapshot();
	} );

	it( 'renders a checked checkbox and a confirmation message when toggled on', async () => {
		const user = userEvent.setup();

		const { container } = render(
			<EnableCustomFieldsOption areCustomFieldsEnabled={ false } />
		);

		await user.click( screen.getByRole( 'checkbox' ) );

		expect( container ).toMatchSnapshot();
	} );
} );

describe( 'CustomFieldsConfirmation', () => {
	it( 'submits the toggle-custom-fields-form', async () => {
		const user = userEvent.setup();
		const submit = jest.fn();
		const setAttribute = jest.fn();
		const getElementById = jest
			.spyOn( document, 'getElementById' )
			.mockImplementation( () => ( {
				submit,
				querySelector: () => ( { setAttribute } ),
			} ) );
		render( <CustomFieldsConfirmation /> );

		await user.click( screen.getByRole( 'button' ) );

		expect( getElementById ).toHaveBeenCalledWith(
			'toggle-custom-fields-form'
		);
		expect( setAttribute ).toHaveBeenCalledWith(
			'value',
			'/' // This is the path returned by getPathAndQueryString.
		);
		expect( submit ).toHaveBeenCalled();

		getElementById.mockRestore();
	} );
} );
