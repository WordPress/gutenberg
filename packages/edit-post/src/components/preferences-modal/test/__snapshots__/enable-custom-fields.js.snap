// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnableCustomFieldsOption renders a checked checkbox and a confirmation message when toggled on 1`] = `
.emotion-0 {
  font-family: -apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen-Sans,Ubuntu,Cantarell,'Helvetica Neue',sans-serif;
  font-size: 13px;
  box-sizing: border-box;
}

.emotion-0 *,
.emotion-0 *::before,
.emotion-0 *::after {
  box-sizing: inherit;
}

.components-panel__row .emotion-2 {
  margin-bottom: inherit;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  gap: calc(4px * 2);
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  width: 100%;
}

.emotion-4>* {
  min-width: 0;
}

.emotion-6 {
  display: block;
  max-height: 100%;
  max-width: 100%;
  min-height: 0;
  min-width: 0;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

<div>
  <div
    class="preference-base-option"
  >
    <div
      class="components-base-control components-toggle-control emotion-0 emotion-1"
    >
      <div
        class="components-base-control__field emotion-2 emotion-3"
      >
        <div
          class="components-flex components-h-stack emotion-4 emotion-5"
          data-wp-c16t="true"
          data-wp-component="HStack"
        >
          <span
            class="components-form-toggle is-checked"
          >
            <input
              class="components-form-toggle__input"
              id="inspector-toggle-control-3"
              type="checkbox"
            />
            <span
              class="components-form-toggle__track"
            />
            <span
              class="components-form-toggle__thumb"
            />
          </span>
          <label
            class="components-flex-item components-flex-block components-toggle-control__label emotion-6 emotion-5"
            data-wp-c16t="true"
            data-wp-component="FlexBlock"
            for="inspector-toggle-control-3"
          />
        </div>
      </div>
    </div>
    <p
      class="edit-post-preferences-modal__custom-fields-confirmation-message"
    >
      A page reload is required for this change. Make sure your content is saved before reloading.
    </p>
    <button
      class="components-button is-next-40px-default-size is-secondary"
      type="button"
    >
      Show & Reload Page
    </button>
  </div>
</div>
`;

exports[`EnableCustomFieldsOption renders a checked checkbox when custom fields are enabled 1`] = `
.emotion-0 {
  font-family: -apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen-Sans,Ubuntu,Cantarell,'Helvetica Neue',sans-serif;
  font-size: 13px;
  box-sizing: border-box;
}

.emotion-0 *,
.emotion-0 *::before,
.emotion-0 *::after {
  box-sizing: inherit;
}

.components-panel__row .emotion-2 {
  margin-bottom: inherit;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  gap: calc(4px * 2);
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  width: 100%;
}

.emotion-4>* {
  min-width: 0;
}

.emotion-6 {
  display: block;
  max-height: 100%;
  max-width: 100%;
  min-height: 0;
  min-width: 0;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

<div>
  <div
    class="preference-base-option"
  >
    <div
      class="components-base-control components-toggle-control emotion-0 emotion-1"
    >
      <div
        class="components-base-control__field emotion-2 emotion-3"
      >
        <div
          class="components-flex components-h-stack emotion-4 emotion-5"
          data-wp-c16t="true"
          data-wp-component="HStack"
        >
          <span
            class="components-form-toggle is-checked"
          >
            <input
              checked=""
              class="components-form-toggle__input"
              id="inspector-toggle-control-0"
              type="checkbox"
            />
            <span
              class="components-form-toggle__track"
            />
            <span
              class="components-form-toggle__thumb"
            />
          </span>
          <label
            class="components-flex-item components-flex-block components-toggle-control__label emotion-6 emotion-5"
            data-wp-c16t="true"
            data-wp-component="FlexBlock"
            for="inspector-toggle-control-0"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`EnableCustomFieldsOption renders an unchecked checkbox and a confirmation message when toggled off 1`] = `
.emotion-0 {
  font-family: -apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen-Sans,Ubuntu,Cantarell,'Helvetica Neue',sans-serif;
  font-size: 13px;
  box-sizing: border-box;
}

.emotion-0 *,
.emotion-0 *::before,
.emotion-0 *::after {
  box-sizing: inherit;
}

.components-panel__row .emotion-2 {
  margin-bottom: inherit;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  gap: calc(4px * 2);
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  width: 100%;
}

.emotion-4>* {
  min-width: 0;
}

.emotion-6 {
  display: block;
  max-height: 100%;
  max-width: 100%;
  min-height: 0;
  min-width: 0;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

<div>
  <div
    class="preference-base-option"
  >
    <div
      class="components-base-control components-toggle-control emotion-0 emotion-1"
    >
      <div
        class="components-base-control__field emotion-2 emotion-3"
      >
        <div
          class="components-flex components-h-stack emotion-4 emotion-5"
          data-wp-c16t="true"
          data-wp-component="HStack"
        >
          <span
            class="components-form-toggle"
          >
            <input
              checked=""
              class="components-form-toggle__input"
              id="inspector-toggle-control-2"
              type="checkbox"
            />
            <span
              class="components-form-toggle__track"
            />
            <span
              class="components-form-toggle__thumb"
            />
          </span>
          <label
            class="components-flex-item components-flex-block components-toggle-control__label emotion-6 emotion-5"
            data-wp-c16t="true"
            data-wp-component="FlexBlock"
            for="inspector-toggle-control-2"
          />
        </div>
      </div>
    </div>
    <p
      class="edit-post-preferences-modal__custom-fields-confirmation-message"
    >
      A page reload is required for this change. Make sure your content is saved before reloading.
    </p>
    <button
      class="components-button is-next-40px-default-size is-secondary"
      type="button"
    >
      Hide & Reload Page
    </button>
  </div>
</div>
`;

exports[`EnableCustomFieldsOption renders an unchecked checkbox when custom fields are disabled 1`] = `
.emotion-0 {
  font-family: -apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen-Sans,Ubuntu,Cantarell,'Helvetica Neue',sans-serif;
  font-size: 13px;
  box-sizing: border-box;
}

.emotion-0 *,
.emotion-0 *::before,
.emotion-0 *::after {
  box-sizing: inherit;
}

.components-panel__row .emotion-2 {
  margin-bottom: inherit;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  gap: calc(4px * 2);
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  width: 100%;
}

.emotion-4>* {
  min-width: 0;
}

.emotion-6 {
  display: block;
  max-height: 100%;
  max-width: 100%;
  min-height: 0;
  min-width: 0;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

<div>
  <div
    class="preference-base-option"
  >
    <div
      class="components-base-control components-toggle-control emotion-0 emotion-1"
    >
      <div
        class="components-base-control__field emotion-2 emotion-3"
      >
        <div
          class="components-flex components-h-stack emotion-4 emotion-5"
          data-wp-c16t="true"
          data-wp-component="HStack"
        >
          <span
            class="components-form-toggle"
          >
            <input
              class="components-form-toggle__input"
              id="inspector-toggle-control-1"
              type="checkbox"
            />
            <span
              class="components-form-toggle__track"
            />
            <span
              class="components-form-toggle__thumb"
            />
          </span>
          <label
            class="components-flex-item components-flex-block components-toggle-control__label emotion-6 emotion-5"
            data-wp-c16t="true"
            data-wp-component="FlexBlock"
            for="inspector-toggle-control-1"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
